### PR TITLE
Fix record

### PR DESCRIPTION
**Pull request by NathanSpeidel**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: src/main/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverter.java [hunk 1]
```diff
@@ -174,7 +177,8 @@ private Serializable fixUnion(List<Schema> types, Object value) {
             // Use the first non null type that works
             try {
                 fixed = fix(schema, value);
-            } catch (Exception ignored) {
+            } catch (Exception e) {
+                log.error(ExceptionUtils.getStackTrace(e));
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverter.java:181]: Just do log.error(e). The trace is there.

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java [hunk 2]
```diff
@@ -37,26 +37,50 @@
...
-                                .name("stringField").type().optional().stringType()
-                                .endRecord()
+                            .enumeration("enumeration")
+                                .symbols("one", "two", "three")
                             .endUnion()
                             .nullDefault()
                         .endRecord();
+
+        RECORD_SCHEMA = SchemaBuilder.record("recordOfString")
+                    .fields()
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:80]: Indentation like SCHEMA above

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java [hunk 3]
```diff
@@ -381,6 +423,48 @@ public void testMapOfMapStringFixing() throws Exception {
...
+    @Test
+    public void testMapOfRecordOfStringFixing() throws Exception {
+        Map<String, GenericRecord> mapOfRecordOfString = new HashMap<>();
+        GenericRecord innerRecord = make(RECORD_SCHEMA, new Field("firstField", "firstValue"), new Field("secondField", "secondValue"));
+        mapOfRecordOfString.put("innerRecordInMap", innerRecord);
+        GenericRecord input = make(new Field("mapOfRecordOfString", mapOfRecordOfString));
+        BulletRecord actual = fixingConverter().convert(input);
+
+        BulletRecord expected = new TypedAvroBulletRecordProvider().getInstance();
+        HashMap<String, HashMap<String, String>> expectedOuterMap = new HashMap<>();
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:435]: Use Maps for these

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java [hunk 4]
```diff
@@ -356,6 +384,20 @@ public void testMapStringFixing() throws Exception {
...
     }
 
+    @Test
+    public void testRecordStringFixing() throws Exception {
+        GenericRecord record = make(RECORD_SCHEMA, new Field("firstField", "firstValue"), new Field("secondField", "secondValue"));
+        GenericRecord input = make(new Field("recordOfString", record));
+        BulletRecord actual = fixingConverter().convert(input);
+
+        BulletRecord expected = new TypedAvroBulletRecordProvider().getInstance();
+        HashMap<String, String> expectedMap = new HashMap<>();
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:394]: Use Map

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java [hunk 5]
```diff
@@ -381,6 +423,48 @@ public void testMapOfMapStringFixing() throws Exception {
...
+        expectedInnerMap.put("secondField", "secondValue");
+        expectedOuterMap.put("innerRecordInMap", expectedInnerMap);
+        expected.typedSet("mapOfRecordOfString", new TypedObject(Type.STRING_MAP_MAP, expectedOuterMap));
+        Assert.assertEquals(actual, expected);
+    }
+
+    @Test
+    public void testRecordOfRecordStringFixing() throws Exception {
+        Schema topMapSchema = SchemaBuilder.record("recordOfRecordOfString").fields()
+                    .name("firstField").type().optional().record("innerRecordInRecord").fields()
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:447]: Indentation here

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java [hunk 6]
```diff
@@ -401,7 +485,28 @@ public void testListStringFixing() throws Exception {
...
+        List<Object> stringRecordList = new ArrayList<>();
+        stringRecordList.add(innerRecord);
+        stringRecordList.add(null);
+
+        GenericRecord input = make(new Field("listOfRecordOfString", stringRecordList));
+        BulletRecord actual = fixingConverter().convert(input);
+
+        BulletRecord expected = new TypedAvroBulletRecordProvider().getInstance();
+        ArrayList<Map<String, String>> expectedStringMapList = new ArrayList<>();
+        HashMap<String, String> expectedStringMap = new HashMap<>();
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:499]: Use List and Map

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java (hunk 2 above)
Comment (NathanSpeidel) [reply src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:80]: This is indented like SCHEMA above.

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java (hunk 4 above)
Comment (NathanSpeidel) [reply src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:394]: Need to be serializable.

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java (hunk 3 above)
Comment (NathanSpeidel) [reply src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:435]: They need to be serializable.

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java (hunk 6 above)
Comment (NathanSpeidel) [reply src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:499]: Need to be serializable.

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java (hunk 2 above)
Comment (akshaisarma) [reply src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:80]: Not the indentation for fields and then subsequent ones

File: src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java [hunk 7]
```diff
@@ -381,6 +423,48 @@ public void testMapOfMapStringFixing() throws Exception {
...
+    }
+
+    @Test
+    public void testRecordOfRecordStringFixing() throws Exception {
+        Schema topMapSchema = SchemaBuilder.record("recordOfRecordOfString").fields()
+                .name("firstField").type().optional().record("innerRecordInRecord").fields()
+                    .name("firstField").type().nullable().stringType().noDefault()
+                    .name("secondField").type().nullable().stringType().noDefault()
+                    .endRecord()
+                .endRecord();
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/dsl/converter/AvroBulletRecordConverterTest.java:451]: Indentation

**Review by akshaisarma — APPROVED** [review]

**Pull request merged**